### PR TITLE
gnrc_ipv6_nib: allow for configuration of static link-local addresses

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This feature was removed without replacment in [71a7dbf91856d](https://github.com/RIOT-OS/RIOT/commit/71a7dbf91856dc221a5615e29587c6b8309b08d2#diff-f9bf85cb337616822921e4230ede1572L857-L869), but it can
still be useful.

### Testing procedure
Uncomment the definition for `GNRC_IPV6_STATIC_LLADDR` in `examples/gnrc_networking/Makefile`

```diff
diff --git a/examples/gnrc_networking/Makefile b/examples/gnrc_networking/Makefile
index 0ae56343f9..6628e113ff 100644
--- a/examples/gnrc_networking/Makefile
+++ b/examples/gnrc_networking/Makefile
@@ -39,8 +39,8 @@ DEVELHELP ?= 1
 # Uncomment the following 2 lines to specify static link lokal IPv6 address
 # this might be useful for testing, in cases where you cannot or do not want to
 # run a shell with ifconfig to get the real link lokal address.
-#IPV6_STATIC_LLADDR ?= '"fe80::cafe:cafe:cafe:1"'
-#CFLAGS += -DGNRC_IPV6_STATIC_LLADDR=$(IPV6_STATIC_LLADDR)
+IPV6_STATIC_LLADDR ?= '"fe80::cafe:cafe:cafe:1"'
+CFLAGS += -DGNRC_IPV6_STATIC_LLADDR=$(IPV6_STATIC_LLADDR)
 
 # Uncomment this to join RPL DODAGs even if DIOs do not contain
 # DODAG Configuration Options (see the doc for more info)
```

Compile and flash to a board of your choice

```sh
make -C examples/gnrc_networking flash term
```

All interfaces of the node should now have two link-local addresses: One starting with the prefix `fe80::cafe:cafe:cafe:` and the one it has without the patch.

The application should also still compile without the patch.

### Issues/PRs references
Closes https://github.com/RIOT-OS/RIOT/pull/13075

### CODE REVIEW

**Comment by fjmolinas** [comment]
Tested:

```
2020-01-14 15:31:58,946 # Iface  7  HWaddr: 11:15  Channel: 26  Page: 0  NID: 0x23
2020-01-14 15:31:58,946 #           Long HWaddr: 06:BC:FB:65:10:6B:11:15 
2020-01-14 15:31:58,957 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-01-14 15:31:58,958 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2020-01-14 15:31:58,959 #           6LO  IPHC  
2020-01-14 15:31:58,959 #           Source address length: 8
2020-01-14 15:31:58,960 #           Link type: wireless
2020-01-14 15:31:58,961 #           inet6 addr: fe80::cafe:cafe:cafe:8  scope: link  VAL
2020-01-14 15:31:58,962 #           inet6 addr: fe80::4bc:fb65:106b:1115  scope: link  VAL
2020-01-14 15:31:58,962 #           inet6 group: ff02::2
2020-01-14 15:31:58,963 #           inet6 group: ff02::1
2020-01-14 15:31:58,964 #           inet6 group: ff02::1:fffe:8
2020-01-14 15:31:58,964 #           inet6 group: ff02::1:ff6b:1115
2020-01-14 15:31:58,972 #           inet6 group: ff02::1a
2020-01-14 15:31:58,972 #           
2020-01-14 15:31:58,973 #           Statistics for Layer 2
2020-01-14 15:31:58,974 #             RX packets 0  bytes 0
2020-01-14 15:31:58,976 #             TX packets 3 (Multicast: 3)  bytes 139
2020-01-14 15:31:58,976 #             TX succeeded 3 errors 0
2020-01-14 15:31:58,977 #           Statistics for IPv6
2020-01-14 15:31:58,978 #             RX packets 0  bytes 0
2020-01-14 15:31:58,980 #             TX packets 3 (Multicast: 3)  bytes 178
2020-01-14 15:31:58,981 #             TX succeeded 3 errors 0
2020-01-14 15:31:58,981 #
```

**Comment by fjmolinas** [comment]
Murdock will verify it compiles, @leandrolanzieri do you agree with this one superseding #13075?

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 1]
```diff
@@ -16,6 +16,7 @@
 #include <errno.h>
 #include <stdbool.h>
 
+#include "kernel_defines.h"
```
Comment (leandrolanzieri) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c]: This is not needed right?

**Comment by leandrolanzieri** [comment]
> Murdock will verify it compiles, @leandrolanzieri do you agree with this one superseding #13075?

Yes sure, better to have the feature in!

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: Oh yeah, was from a previous iteration where I thought I could use `IS_ACTIVE()` (which apparently doesn't work with strings)

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: Amended that change. I hope this was okay.

Comment (leandrolanzieri) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: Yeah, I was thinking about that. One way I can think of is having a `PSEUDOMODULE` to enable the feature and a second macro to pass the value. That way you can use C conditionals

**Comment by fjmolinas** [comment]
@leandrolanzieri this seems ready

**Comment by miri64** [comment]
Thanks for the quick review!

**Comment by benpicco** [comment]
The only problem is that the node is not reachable under that address:

**samr21-xpro**
```
2020-02-12 19:13:23,411 # Iface  7  HWaddr: 7D:79  Channel: 26  Page: 0  NID: 0x23
2020-02-12 19:13:23,416 #           Long HWaddr: 5A:06:59:C9:2C:36:7D:79 
2020-02-12 19:13:23,423 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-02-12 19:13:23,430 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2020-02-12 19:13:23,433 #           6LO  IPHC  
2020-02-12 19:13:23,438 #           Source address length: 8
2020-02-12 19:13:23,441 #           Link type: wireless
2020-02-12 19:13:23,445 #           inet6 addr: fe80::cafe:cafe:cafe:8  scope: link  VAL
2020-02-12 19:13:23,449 #           inet6 addr: fe80::5806:59c9:2c36:7d79  scope: link  VAL
2020-02-12 19:13:23,455 #           inet6 group: ff02::2
2020-02-12 19:13:23,456 #           inet6 group: ff02::1
2020-02-12 19:13:23,458 #           inet6 group: ff02::1:fffe:8
2020-02-12 19:13:23,461 #           inet6 group: ff02::1:ff36:7d79
2020-02-12 19:13:23,464 #           inet6 group: ff02::1a
2020-02-12 19:13:23,465 #           
2020-02-12 19:13:23,474 #           Statistics for Layer 2
2020-02-12 19:13:23,477 #             RX packets 15  bytes 697
2020-02-12 19:13:23,479 #             TX packets 17 (Multicast: 13)  bytes 777
2020-02-12 19:13:23,480 #             TX succeeded 17 errors 0
2020-02-12 19:13:23,482 #           Statistics for IPv6
2020-02-12 19:13:23,485 #             RX packets 15  bytes 904
2020-02-12 19:13:23,490 #             TX packets 20 (Multicast: 16)  bytes 1182
2020-02-12 19:13:23,493 #             TX succeeded 20 errors 0
2020-02-12 19:13:23,494 # 
```

**other node**
```
2020-02-12 19:14:15,024 # Iface  8  HWaddr: 4C:D3  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
2020-02-12 19:14:15,027 #            high data rate: 0 
2020-02-12 19:14:15,030 #           Long HWaddr: FA:73:82:30:0D:3E:1B:8C 
2020-02-12 19:14:15,038 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-02-12 19:14:15,043 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2020-02-12 19:14:15,046 #           6LO  IPHC  
2020-02-12 19:14:15,048 #           Source address length: 8
2020-02-12 19:14:15,050 #           Link type: wireless
2020-02-12 19:14:15,055 #           inet6 addr: fe80::f873:8230:d3e:1b8c  scope: link  VAL
2020-02-12 19:14:15,058 #           inet6 group: ff02::2
2020-02-12 19:14:15,064 #           inet6 group: ff02::1
2020-02-12 19:14:15,066 #           inet6 group: ff02::1:ff3e:1b8c
2020-02-12 19:14:15,067 #           
2020-02-12 19:14:15,069 #           Statistics for Layer 2
2020-02-12 19:14:15,069 #             RX packets 0  bytes 0
2020-02-12 19:14:15,074 #             TX packets 2 (Multicast: 2)  bytes 86
2020-02-12 19:14:15,076 #             TX succeeded 2 errors 0
2020-02-12 19:14:15,079 #           Statistics for IPv6
2020-02-12 19:14:15,084 #             RX packets 0  bytes 0
2020-02-12 19:14:15,086 #             TX packets 2 (Multicast: 2)  bytes 128
2020-02-12 19:14:15,090 #             TX succeeded 2 errors 0

2020-02-12 19:14:41,062 #  ping6 fe80::cafe:cafe:cafe:8%8
2020-02-12 19:14:44,078 # 
2020-02-12 19:14:44,081 # --- fe80::cafe:cafe:cafe:8 PING statistics ---
2020-02-12 19:14:44,086 # 3 packets transmitted, 0 packets received, 100% packet loss
2020-02-12 19:14:54,129 #  ping6 fe80::5806:59c9:2c36:7d79%8
2020-02-12 19:14:54,145 # 12 bytes from fe80::5806:59c9:2c36:7d79: icmp_seq=0 ttl=64 rssi=-69 dBm time=8.269 ms
2020-02-12 19:14:55,146 # 12 bytes from fe80::5806:59c9:2c36:7d79: icmp_seq=1 ttl=64 rssi=-69 dBm time=6.983 ms
2020-02-12 19:14:56,146 # 12 bytes from fe80::5806:59c9:2c36:7d79: icmp_seq=2 ttl=64 rssi=-69 dBm time=6.998 ms
2020-02-12 19:14:56,146 # 
2020-02-12 19:14:56,152 # --- fe80::5806:59c9:2c36:7d79 PING statistics ---
2020-02-12 19:14:56,157 # 3 packets transmitted, 3 packets received, 0% packet loss
2020-02-12 19:14:56,160 # round-trip min/avg/max = 6.983/7.416/8.269 ms
```

**Comment by miri64** [comment]
> The only problem is that the node is not reachable under that address:
> 
> **samr21-xpro**
> 
> ```
> 2020-02-12 19:13:23,411 # Iface  7  HWaddr: 7D:79  Channel: 26  Page: 0  NID: 0x23
> 2020-02-12 19:13:23,416 #           Long HWaddr: 5A:06:59:C9:2C:36:7D:79 
> 2020-02-12 19:13:23,423 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
> 2020-02-12 19:13:23,430 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
> 2020-02-12 19:13:23,433 #           6LO  IPHC  
> 2020-02-12 19:13:23,438 #           Source address length: 8
> 2020-02-12 19:13:23,441 #           Link type: wireless
> 2020-02-12 19:13:23,445 #           inet6 addr: fe80::cafe:cafe:cafe:8  scope: link  VAL
> 2020-02-12 19:13:23,449 #           inet6 addr: fe80::5806:59c9:2c36:7d79  scope: link  VAL
> 2020-02-12 19:13:23,455 #           inet6 group: ff02::2
> 2020-02-12 19:13:23,456 #           inet6 group: ff02::1
> 2020-02-12 19:13:23,458 #           inet6 group: ff02::1:fffe:8
> 2020-02-12 19:13:23,461 #           inet6 group: ff02::1:ff36:7d79
> 2020-02-12 19:13:23,464 #           inet6 group: ff02::1a
> 2020-02-12 19:13:23,465 #           
> 2020-02-12 19:13:23,474 #           Statistics for Layer 2
> 2020-02-12 19:13:23,477 #             RX packets 15  bytes 697
> 2020-02-12 19:13:23,479 #             TX packets 17 (Multicast: 13)  bytes 777
> 2020-02-12 19:13:23,480 #             TX succeeded 17 errors 0
> 2020-02-12 19:13:23,482 #           Statistics for IPv6
> 2020-02-12 19:13:23,485 #             RX packets 15  bytes 904
> 2020-02-12 19:13:23,490 #             TX packets 20 (Multicast: 16)  bytes 1182
> 2020-02-12 19:13:23,493 #             TX succeeded 20 errors 0
> 2020-02-12 19:13:23,494 # 
> ```
> 
> **other node**
> 
> ```
> 2020-02-12 19:14:15,024 # Iface  8  HWaddr: 4C:D3  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
> 2020-02-12 19:14:15,027 #            high data rate: 0 
> 2020-02-12 19:14:15,030 #           Long HWaddr: FA:73:82:30:0D:3E:1B:8C 
> 2020-02-12 19:14:15,038 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
> 2020-02-12 19:14:15,043 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
> 2020-02-12 19:14:15,046 #           6LO  IPHC  
> 2020-02-12 19:14:15,048 #           Source address length: 8
> 2020-02-12 19:14:15,050 #           Link type: wireless
> 2020-02-12 19:14:15,055 #           inet6 addr: fe80::f873:8230:d3e:1b8c  scope: link  VAL
> 2020-02-12 19:14:15,058 #           inet6 group: ff02::2
> 2020-02-12 19:14:15,064 #           inet6 group: ff02::1
> 2020-02-12 19:14:15,066 #           inet6 group: ff02::1:ff3e:1b8c
> 2020-02-12 19:14:15,067 #           
> 2020-02-12 19:14:15,069 #           Statistics for Layer 2
> 2020-02-12 19:14:15,069 #             RX packets 0  bytes 0
> 2020-02-12 19:14:15,074 #             TX packets 2 (Multicast: 2)  bytes 86
> 2020-02-12 19:14:15,076 #             TX succeeded 2 errors 0
> 2020-02-12 19:14:15,079 #           Statistics for IPv6
> 2020-02-12 19:14:15,084 #             RX packets 0  bytes 0
> 2020-02-12 19:14:15,086 #             TX packets 2 (Multicast: 2)  bytes 128
> 2020-02-12 19:14:15,090 #             TX succeeded 2 errors 0
> 
> 2020-02-12 19:14:41,062 #  ping6 fe80::cafe:cafe:cafe:8%8
> 2020-02-12 19:14:44,078 # 
> 2020-02-12 19:14:44,081 # --- fe80::cafe:cafe:cafe:8 PING statistics ---
> 2020-02-12 19:14:44,086 # 3 packets transmitted, 0 packets received, 100% packet loss
> 2020-02-12 19:14:54,129 #  ping6 fe80::5806:59c9:2c36:7d79%8
> 2020-02-12 19:14:54,145 # 12 bytes from fe80::5806:59c9:2c36:7d79: icmp_seq=0 ttl=64 rssi=-69 dBm time=8.269 ms
> 2020-02-12 19:14:55,146 # 12 bytes from fe80::5806:59c9:2c36:7d79: icmp_seq=1 ttl=64 rssi=-69 dBm time=6.983 ms
> 2020-02-12 19:14:56,146 # 12 bytes from fe80::5806:59c9:2c36:7d79: icmp_seq=2 ttl=64 rssi=-69 dBm time=6.998 ms
> 2020-02-12 19:14:56,146 # 
> 2020-02-12 19:14:56,152 # --- fe80::5806:59c9:2c36:7d79 PING statistics ---
> 2020-02-12 19:14:56,157 # 3 packets transmitted, 3 packets received, 0% packet loss
> 2020-02-12 19:14:56,160 # round-trip min/avg/max = 6.983/7.416/8.269 ms
> ```

If you configure your neighbor cache correctly it is. With 6Lo-ND address resolution won't work with that address. See https://tools.ietf.org/html/rfc6775#section-5.6

**Comment by benpicco** [comment]
> If you configure your neighbor cache correctly it is.

Any pointers on how to do that? The comment in the Makefile says all I need to do is uncomment those two lines.

**Comment by miri64** [comment]
Either by deactivating 6Lo-ND (look at what dependencies the `gnrc_ipv6..._default` pulls in and replace `gnrc_ipv6..._default` with them but skip the ones containing `nib_6l` or `sixlowpan_nd`; attention this also deactivates multihop prefix dissemination [can be re-introduced with RPL] and compression context dissemination) or by using the `nib neigh` (`ip -6 neigh` on Linux) command to update the neighbor cache manually. I think the comment just takes classic ND into account (e.g. over Ethernet or WiFi).

**Comment by benpicco** [comment]
Hm, could all that be avoided by using a global or site-local address?
A simple `ifconfig 8 add fec0::1` doesn't seem to cut it though - can't ping that address from other nodes. I thought the nodes *would* do neighbor discovery for non-link-local addresses?

**Comment by miri64** [comment]
I'm starting to wonder what you are trying to do??!? This feature is purely meant for debugging purposes (or when for whatever reason the auto-configured link-local or dynamically configured address is not working out).

> Hm, could all that be avoided by using a global or site-local address?
> A simple `ifconfig 8 add fec0::1` doesn't seem to cut it though - can't ping that address from other nodes. I thought the nodes _would_ do neighbor discovery for non-link-local addresses?

That seems unrelated to this PR.

**Comment by miri64** [comment]
In any way: using a link-local address _not_ based on a 6LoWPAN interface's EUI-64, [breaks the spec in RFC6775](https://tools.ietf.org/html/rfc6775#section-5.2). If you decide to steer of standard, you have make due with the consequences.

**Comment by benpicco** [comment]
I'm trying to set up an automatic test that gets run on two random nodes. (we're building a small testbed, you say how many nodes you want and you get assigned a random set once enough nodes are available)
So I want to configure the two nodes with a known IP to run some simple ping-pong tests between them.
Doesn't have to be a link-local address, but that's what was already provided by the `gnrc_networking` Makefile.

**Comment by benpicco** [comment]
> using a link-local address not based on a 6LoWPAN interface's EUI-64, breaks the spec in RFC6775.

I get that, but the consequence would be

- a 6lo interface can only have one link-local address (or however many frame filters the HW supports)
- changing the link-local address must also update the hardware address

Both seem like reasonable requirements - patches welcome?

**Comment by benpicco** [comment]
> In any way: using a link-local address _not_ based on a 6LoWPAN interface's EUI-64, [breaks the spec in RFC6775](https://tools.ietf.org/html/rfc6775#section-5.2).

So should we do

```patch
--- a/sys/net/gnrc/network_layer/ipv6/nib/nib.c
+++ b/sys/net/gnrc/network_layer/ipv6/nib/nib.c
@@ -143,7 +143,10 @@ void gnrc_ipv6_nib_iface_up(gnrc_netif_t *netif)
     _init_iface_arsm(netif);
     netif->ipv6.rs_sent = 0;
     netif->ipv6.na_sent = 0;
-    _add_static_lladdr(netif);
+
+    if (!gnrc_netif_is_6ln(netif)) {
+        _add_static_lladdr(netif);
+    }
     _auto_configure_addr(netif, &ipv6_addr_link_local_prefix, 64U);
     if (_should_search_rtr(netif)) {
         uint32_t next_rs_time = random_uint32_range(0, NDP_MAX_RS_MS_DELAY);
```

**Pull request merged**